autorest/sender.go

```diff
@@ -233,30 +241,48 @@ func DoRetryForStatusCodes(attempts int, backoff time.Duration, codes ...int) Se
 					return resp, err
 				}
 				delayed := DelayWithRetryAfter(resp, r.Context().Done())
-				if !delayed && !DelayForBackoff(backoff, attempt, r.Context().Done()) {
+				// default delay cap of 0 means no cap
+				delayCap := time.Duration(0)
+				// if this was a 429 set the delay cap as specified.
+				// applicable only in the absense of a retry-after header.
+				if resp != nil && resp.StatusCode == http.StatusTooManyRequests {
+					delayCap = Max429Delay
+				}
+				if !delayed && !delayForBackoffWithCap(backoff, delayCap, delayCount, r.Context().Done()) {
 					return resp, r.Context().Err()
 				}
-				// don't count a 429 against the number of attempts
-				// so that we continue to retry until it succeeds
-				if resp == nil || resp.StatusCode != http.StatusTooManyRequests {
+				// when Count429AsRetry == false don't count a 429 against the number
+				// of attempts so that we continue to retry until it succeeds
+				if Count429AsRetry || (resp == nil || resp.StatusCode != http.StatusTooManyRequests) {
 					attempt++
 				}
+				// delay count is tracked separately from attempts to
+				// ensure that 429 participates in exponential back-off
+				delayCount++
 			}
 			return resp, err
 		})
 	}
 }
 
-// DelayWithRetryAfter invokes time.After for the duration specified in the "Retry-After" header in
-// responses with status code 429
+// DelayWithRetryAfter invokes time.After for the duration specified in the "Retry-After" header.
+// The value of Retry-After can be either the number of seconds or a date in RFC1123 format.
+// The function returns true after successfully waiting for the specified duration.  If there is
+// no Retry-After header or the wait is cancelled the return value is false.
 func DelayWithRetryAfter(resp *http.Response, cancel <-chan struct{}) bool {
 	if resp == nil {
 		return false
 	}
-	retryAfter, _ := strconv.Atoi(resp.Header.Get("Retry-After"))
-	if resp.StatusCode == http.StatusTooManyRequests && retryAfter > 0 {
+	var dur time.Duration
+	ra := resp.Header.Get("Retry-After")
+	if retryAfter, _ := strconv.Atoi(ra); retryAfter > 0 {
+		dur = time.Duration(retryAfter) * time.Second
+	} else if t, err := time.Parse(time.RFC1123, ra); err == nil {
```

Comment:
IF err != nil  then we return false and there will be no delay?

Reply:
The exponential back-off delay will be used instead.